utils/rand/rand.go

```diff
@@ -0,0 +1,169 @@
+package rand
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// This package is a wrppaer around true RNG crypto/rand.
+// It implements useful tools using the true RNG and that
+// are not exported by the crypto/rand package.
+// This package does not implement any determinstic RNG (Pseudo RNG)
+// unlike the package flow-go/crypto/random.
+
+var randFailure = errors.New("crypto/rand failed")
+
+// returns a random uint64
+func Uint64() (uint64, error) {
+	buffer := make([]byte, 8) // TODO: declare as a global variable and add a lock?
+	if _, err := rand.Read(buffer); err != nil {
+		return 0, randFailure
+	}
+	r := binary.LittleEndian.Uint64(buffer)
+	return r, nil
+}
+
+// returns a random uint64 strictly less than n
+// errors if n==0
+func Uint64n(n uint64) (uint64, error) {
+	if n == 0 {
+		return 0, fmt.Errorf("n should be strictly positive, got %d", n)
+	}
+	// the max returned random is n-1 > 0
+	max := n - 1
+	// count the bytes size of max
+	size := 0
+	for tmp := max; tmp != 0; tmp >>= 8 {
+		size++
+	}
+	buffer := make([]byte, 8) // TODO: declare as a global variable and add a lock?
+	// get the bit size of max
+	mask := uint64(0)
+	for max&mask != max {
+		mask = (mask << 1) | 1
+	}
+
+	// Using 64 bits of random and reducing modulo n does not guarantee a high uniformity
+	// of the result.
+	// For a better uniformity, loop till a sample is less or equal to `max`.
+	// This means the function might take longer time to output a random.
+	// Using the size of `max` in bits helps the loop end earlier (the algo stops after one loop
+	// with more than 50%)
+	// a different approach would be to pull at least 128 bits from the random source
+	// and use big number modular reduction by `n`.
+	random := n
+	for random > max {
+		if _, err := rand.Read(buffer[:size]); err != nil {
+			return 0, randFailure
+		}
+		random = binary.LittleEndian.Uint64(buffer)
+		random &= mask // adjust to the size of max in bits
+	}
+
+	return random, nil
+}
+
+// returns a random uint32
+func Uint32() (uint32, error) {
+	// for 64-bits machines, doing 64 bits operations and then casting
+	// should be faster than dealing with 32 bits operations
+	r, err := Uint64()
+	return uint32(r), err
+}
+
+// returns a random uint32 strictly less than n
+// errors if n==0
+func Uint32n(n uint32) (uint32, error) {
```

Comment:
maybe just use Uint64n and cast down?  we're probably not using this in a tight loop, so perf don't matter too much

Reply:
That's true, the only reason here is to optimize for performance as I'm not sure how this tool can be used. 
Now that the more optimized version is available, maybe we could keep it in favour of code reuse?

Reply:
in that case, refactor the code to share the same generic helper function

Reply:
The helper function would need to use internal types of either `uint32` or `uint64`, so it won't have much in common other than the logic. I may just use your initial suggestion and give up on performance.

Reply:
https://github.com/onflow/flow-go/pull/4062/commits/6197afa1fd754265630e573dfffec94baeca25cd